File: web_source_code_visualization/scripts/analyze.py
import ast
import json
import os
import sys
import glob

def analyze_flask(target_dir):
    routes = []
    
    # Python files
    files = glob.glob(os.path.join(target_dir, '**/*.py'), recursive=True)
    
    for file_path in files:
        if 'venv' in file_path or '__pycache__' in file_path:
            continue
            
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                code = f.read()
                
            tree = ast.parse(code, filename=file_path)
            
            for node in ast.walk(tree):
                if isinstance(node, ast.FunctionDef):
                    for decorator in node.decorator_list:
                        # Check for @app.route or @blueprint.route
                        if isinstance(decorator, ast.Call):
                            func = decorator.func
                            is_route = False
                            
                            if isinstance(func, ast.Attribute) and func.attr == 'route':
                                is_route = True
                            
                            if is_route:
                                # Extract path
                                route_path = '/'
                                if decorator.args and isinstance(decorator.args[0], ast.Constant): # Python 3.8+
                                    route_path = decorator.args[0].value
                                elif decorator.args and isinstance(decorator.args[0], ast.Str): # Python < 3.8
                                    route_path = decorator.args[0].s
                                
                                # Extract methods
                                methods = ['GET'] # Default
                                for keyword in decorator.keywords:
                                    if keyword.arg == 'methods' and isinstance(keyword.value, ast.List):
                                        methods = []
                                        for elt in keyword.value.elts:
                                            if isinstance(elt, ast.Constant):
                                                methods.append(elt.value)
                                            elif isinstance(elt, ast.Str):
                                                methods.append(elt.s)
                                
                                for method in methods:
                                    routes.push({ # Logic error in python, fixing below
                                       # Actually this is python list, use append
                                       pass
                                    })
                                    routes.append({
                                        'file': os.path.relpath(file_path, target_dir),
                                        'line': node.lineno,
                                        'method': method.upper(),
                                        'path': route_path,
                                        'type': 'flask',
                                        'framework': 'flask'
                                    })
                                    
        except Exception as e:
            # print(f"Error parsing {file_path}: {e}", file=sys.stderr)
            pass
            
    return routes

if __name__ == "__main__":
    target = sys.argv[1] if len(sys.argv) > 1 else '.'
    result = analyze_flask(target)
    print(json.dumps(result, indent=2))

File: web_source_code_visualization/scripts/analyze.js
const fs = require('fs');
const path = require('path');
const glob = require('glob');
const parser = require('@babel/parser');
const traverse = require('@babel/traverse').default;

// Clean argument path for Windows
let targetDir = process.argv[2] || '.';
targetDir = targetDir.replace(/\\/g, '/'); // Normalize slashes

// Debug to stderr (visible in Next.js console)
console.error(`Analyzing directory: ${targetDir}`);

const files = glob.sync('**/*.{js,ts}', {
    cwd: targetDir,
    ignore: ['**/node_modules/**', '**/dist/**', '**/.next/**'],
    absolute: true
});

console.error(`Found ${files.length} files.`);

const routes = [];

files.forEach(filePath => {
    try {
        const code = fs.readFileSync(filePath, 'utf-8');

        // Parse as unkwown/module to allow both
        const ast = parser.parse(code, {
            sourceType: 'unambiguous',
            plugins: ['typescript', 'jsx']
        });

        traverse(ast, {
            CallExpression({ node }) {
                if (node.callee.type === 'MemberExpression') {
                    const propertyName = node.callee.property.name;
                    // const objectName = node.callee.object.name; // app or router

                    const methods = ['get', 'post', 'put', 'delete', 'patch', 'use'];

                    if (methods.includes(propertyName)) {
                        const args = node.arguments;
                        if (args.length > 0) {
                            let routePath = '/';
                            if (args[0].type === 'StringLiteral') {
                                routePath = args[0].value;
                            } else if (args[0].type === 'TemplateLiteral') {
                                routePath = args[0].quasis.map(q => q.value.raw).join('{param}');
                            } else {
                                return;
                            }

                            // --- Security Analysis Start ---
                            const params = new Set();
                            const sinks = [];
                            let hasSanitization = false;

                            // Helper to scan function body
                            const handler = args.find(arg => arg.type === 'ArrowFunctionExpression' || arg.type === 'FunctionExpression');

                            if (handler && handler.body) {
                                traverse(handler.body, {
                                    noScope: true, // Simple traversal inside handler

                                    // 1. Parameter Extraction (req.body.x, req.query.y)
                                    MemberExpression(path) {
                                        const n = path.node;
                                        if (n.object.type === 'MemberExpression' && n.object.object.name === 'req') {
                                            // req.body.username -> property.name = username
                                            if (['body', 'query', 'params'].includes(n.object.property.name)) {
                                                params.add(`${n.object.property.name}.${n.property.name}`);
                                            }
                                        }
                                        // Destructuring: const { x } = req.body
                                        if (n.object.name === 'req' && ['body', 'query', 'params'].includes(n.property.name)) {
                                            // This is harder to track without full scope analysis, identifying via VariableDeclarator below
                                        }
                                    },

                                    VariableDeclarator(path) {
                                        const n = path.node;
                                        if (n.init && n.init.type === 'MemberExpression' && n.init.object.name === 'req') {
                                            // const { x } = req.body;
                                            if (n.id.type === 'ObjectPattern') {
                                                n.id.properties.forEach(p => {
                                                    params.add(`${n.init.property.name}.${p.key.name}`);
                                                });
                                            }
                                        }
                                    },

                                    // 2. Sink Detection (Danger Patterns)
                                    CallExpression(path) {
                                        const n = path.node;
                                        const callee = n.callee;

                                        // eval(), exec()
                                        if (callee.name === 'eval') sinks.push({ type: 'RCE', detail: 'eval() detected' });

                                        // child_process.exec(cmd)
                                        if (callee.property && ['exec', 'spawn'].includes(callee.property.name)) {
                                            sinks.push({ type: 'RCE', detail: 'Process execution detected' });
                                        }

                                        // db.query(), pool.execute() - Naive SQL heuristic
                                        if (callee.property && ['query', 'execute'].includes(callee.property.name)) {
                                            // Check arguments for string concatenation
                                            const queryArg = n.arguments[0];
                                            if (queryArg && queryArg.type === 'BinaryExpression' && queryArg.operator === '+') {
                                                sinks.push({ type: 'SQLi', detail: 'Potential SQL Injection (String concat)' });
                                            }
                                        }
                                    }
                                }, handler, { visitorKeys: null }); // Need correct visitor context or manual recursion? 
                                // Babel traverse inside traverse can be tricky. Using manual AST walk or scope-less traverse is safer for simple scripts.
                                // Re-implementing simple recursive walk for the handler body to avoid scope issues.
                            }
                            // --- Security Analysis End ---

                            routes.push({
                                file: path.relative(targetDir, filePath),
                                useRelativePath: true, // marker
                                line: node.loc.start.line,
                                method: propertyName.toUpperCase(),
                                path: routePath,
                                type: 'express',
                                framework: 'express',
                                params: Array.from(params),
                                sinks: sinks,
                                riskLevel: sinks.length > 0 ? 'critical' : 'low'
                            });
                        }
                    }
                }
            }
        });
    } catch (e) {
        console.error(`Error parsing ${filePath}:`, e.message);
    }
});

console.log(JSON.stringify(routes, null, 2));
